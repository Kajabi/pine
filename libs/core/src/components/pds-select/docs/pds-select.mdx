import { DocArgsTable, DocCanvas } from '@pine-ds/doc-components';
import { components } from '../../../../dist/docs.json';

# Select

<p>
  The select field, when combined with a label, helper message, and error text, offers a comprehensive control that provides a menu of options for user selection. This ensures
  clarity and guidance for the user, enhancing the overall user experience.
</p>

## Guidelines

### Usage

#### When to use

- When you want to provide a simple list of options for the user to choose from.
- **Example**: Use a select component to choose a state from a short list of U.S. regions.

#### When **NOT** to use

- If your list of options is greater than 10, consider using a custom dropdown component. (WIP)
- **Example**: Use a custom dropdown for selecting multiple filters or choosing a specific product from a comprehensive list. (WIP)

### Accessibility

- The select component comes with a built-in label through the `label` property, it is strongly encouraged that you include it.
- Please provide clear and descriptive error, and helper-message text.
- Selects can have a default option that is not selectable.

## Properties

<DocArgsTable componentName="pds-select" docSource={components} />

### Default

<p>This example shows the default state of the select component with a label and a list of options. The user can choose one of the options from the dropdown menu.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect componentId="pds-select-default-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select component-id="pds-select-default-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select component-id="pds-select-default-example" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

### Disabled

<p>This example demonstrates the select component in a disabled state. The user cannot interact with the dropdown menu or select any options.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect componentId="pds-select-disabled-example" disabled="true" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select component-id="pds-select-disabled-example" disabled="true" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select component-id="pds-select-disabled-example" disabled="true" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

### With Message

<p>This example includes a helper message that provides additional context or instructions for the user. The helper message appears below the select component.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect componentId="pds-select-with-message-example" label="Name" helper-message="This is a helper message to convey extra context for the select">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select component-id="pds-select-with-message-example" label="Name" helper-message="This is a helper message to convey extra context for the select">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select component-id="pds-select-with-message-example" helper-message="This is a helper message to convey extra context for the select" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

### Invalid

<p>This example shows the select component with an error message. The error message provides feedback to the user when there is an issue with their selection.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect componentId="pds-select-invalid-example" label="Name" error-message="This is an error message to convey extra context for the select">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select component-id="pds-select-invalid-example" label="Name" error-message="This is an error message to convey extra context for the select">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select component-id="pds-select-invalid-example" error-message="This is an error message to convey extra context for the select" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

### Autocomplete

<p>This example demonstrates the select component with the autocomplete attribute enabled. The browser may provide autocomplete suggestions based on the user's input.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect autocomplete="on" componentId="pds-select-autocomplete-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select autocomplete="on" component-id="pds-select-autocomplete-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select autocomplete="on" component-id="pds-select-autocomplete-example" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

### Multiple

<p>This example demonstrates the select component with the multiple attribute enabled. The user can select multiple options from the select component.</p>

<DocCanvas
  client:only
  mdxSource={{
    react: `
      <PdsSelect multiple componentId="pds-select-multiple-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </PdsSelect>`,
    webComponent: `
      <pds-select multiple component-id="pds-select-multiple-example" label="Name">
        <option value="paul">Paul McCartney</option>
        <option value="john">John Lennon</option>
        <option value="george">George Harrison</option>
        <option value="ringo">Ringo Starr</option>
      </pds-select>`,
  }}
>
  <div style={{ padding: '20px', width: '100%' }}>
    <pds-select multiple component-id="pds-select-multiple-example" label="Name">
      <option value="paul">Paul McCartney</option>
      <option value="john">John Lennon</option>
      <option value="george">George Harrison</option>
      <option value="ringo">Ringo Starr</option>
    </pds-select>
  </div>
</DocCanvas>

## Additional Resources

<p>[MDN Web Docs: Select](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select)</p>
<p>[MDN Web Docs: HTML autocomplete attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)</p>
