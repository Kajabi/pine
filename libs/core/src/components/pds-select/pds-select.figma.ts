import figma, { html } from '@figma/code-connect/html';

figma.connect('<FIGMA_SELECT>', {
  props: {
    disabled: figma.enum("State", {
      disabled: "true",
    }),
    errorMessage: figma.nestedProps(".error-message", {
      text: figma.string("Error content"),
    }),
    helperMessage: figma.boolean("Help text", {
      true: figma.string("↪️ Message"),
      false: undefined,
    }),
    label: figma.boolean("Label", {
      true: figma.string("↪️ Message"),
      false: undefined,
    }),
    readonly: figma.enum("State", {
      'read only': "true",
      false: undefined,
    }),
    value: figma.nestedProps(".Input text", {
      text: figma.string("Message"),
    }),
  },
  example: (props) => html`\
  <pds-select
    disabled=${props.disabled}
    error-message=${props.errorMessage.text}
    helper-message=${props.helperMessage}
    label=${props.label}
    readonly=${props.readonly}
    value=${props.value.text}
  ></pds-select>`,
});