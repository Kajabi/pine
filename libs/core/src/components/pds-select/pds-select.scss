:host {
  --pds-select-background: var(--pine-color-background-container);
  --pds-select-border: var(--pine-border);
  --pds-select-border-radius: var(--pine-dimension-125);
  --pds-select-border-radius-start-end: var(--pine-dimension-125);
  --pds-select-border-radius-start-start: var(--pine-dimension-125);
  --pds-select-border-radius-end-end: var(--pine-dimension-125);
  --pds-select-border-radius-end-start: var(--pine-dimension-125);
  --pds-select-min-height: 36px;
  --box-shadow-focus: 0 0 0 1px #ffffff, 0 0 0 3px var(--pine-color-focus-ring);
  --box-shadow-focus-error: 0 0 0 1px #ffffff, 0 0 0 3px var(--pine-color-focus-ring-danger);

  .hidden,
  :host([hidden]) {
    display: none;
  }
}

.pds-select {
  display: grid;
  flex-direction: column;
  grid-template-areas:
    'label label'
    'field field'
    'message message';
  grid-template-columns: 1fr minmax(2rem, min-content);
  grid-template-rows: min-content min-content min-content;
  position: relative;
  width: 100%;
}

.pds-select__label-wrapper {
  align-items: center;
  display: flex;
  grid-area: label;
  justify-content: space-between;
  margin-block-end: var(--pine-dimension-2xs);
}

.pds-select__action {
  align-items: center;
  display: flex;
  gap: var(--pine-dimension-xs);
  margin-inline-start: var(--pine-dimension-xs);
}

label {
  margin-block-end: var(--pine-dimension-2xs);
}

// When label is inside wrapper, remove its margin
.pds-select__label-wrapper label {
  margin-block-end: 0;
}

select {
  appearance: none;
  background-color: var(--pds-select-background);
  border: var(--pds-select-border);
  border-radius: var(--pds-select-border-radius);
  border-bottom-left-radius: var(--pds-select-border-radius-end-start, var(--pds-select-border-radius));
  border-bottom-right-radius: var(--pds-select-border-radius-end-end, var(--pds-select-border-radius));
  border-top-left-radius: var(--pds-select-border-radius-start-start, var(--pds-select-border-radius));
  border-top-right-radius: var(--pds-select-border-radius-start-end, var(--pds-select-border-radius));
  font: var(--pine-typography-body);
  grid-area: field;
  letter-spacing: var(--pine-letter-spacing);
  min-height: var(--pds-select-min-height);
  padding: var(--pine-dimension-xs) var(--pine-dimension-150);
  padding-inline-end: var(--pine-dimension-450);
  position: relative;

  &:hover {
    border-color: var(--pine-color-border-hover);
    cursor: pointer;
  }

  &:focus-visible {
    border-color: var(--pine-color-border-active);
    outline: var(--pine-outline-focus);
    outline-offset: var(--pine-border-width);
  }

  &:disabled {
    background-color: var(--pine-color-background-container-disabled);
    border-color: var(--pine-color-border-disabled);
    color: var(--pine-color-text-disabled);
    cursor: not-allowed;
  }

  &:has(~ .pds-select__message .pds-select__error-message) {
    background-color: var(--pine-select-color-background-danger);
    border-color: var(--pine-color-border-danger);

    &:focus-visible {
      outline-color: var(--pine-color-focus-ring-danger);
    }
  }
}

:host(.is-invalid) {
  select {
    background-color: var(--pine-select-color-background-danger);
    border-color: var(--pine-color-border-danger);

    &:focus-visible {
      outline-color: var(--pine-color-focus-ring-danger);
    }
  }
}

.pds-select__error-message,
.pds-select__helper-message {
  font: var(--pine-typography-body-sm-medium);
  margin-block-end: 0;
  margin-block-start: var(--pine-dimension-2xs);
}

.pds-select__error-message {
  align-items: center;
  color: var(--pine-color-text-message-danger);
  display: flex;
  gap: var(--pine-dimension-2xs);
}

.pds-select__message {
  grid-area: message;
}

.pds-select__select-icon {
  align-items: center;
  display: flex;
  grid-column: -1 / -2;
  grid-row: 2 / 3;
  height: var(--pds-select-min-height);
  pointer-events: none;
  position: relative;
  z-index: 1;
}

.visually-hidden {
  border: 0;
  clip: rect(0 0 0 0);
  clip-path: polygon(0 0, 0 0, 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

