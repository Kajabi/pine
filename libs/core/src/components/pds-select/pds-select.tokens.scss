/**
 * Do not edit directly, this file was auto-generated.
 */

:host {
  --pine-select-color-background-danger: var(--pine-color-red-050);
}
