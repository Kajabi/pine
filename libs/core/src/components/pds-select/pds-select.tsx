import { Component, Element, Host, h, Method, Prop, Event, EventEmitter, Listen } from '@stencil/core';
import {
  positionTooltip
} from '../../utils/overlay';
import { assignDescription, messageId } from '../../utils/form';

@Component({
  tag: 'pds-select',
  styleUrl: 'pds-select.scss',
  shadow: true,
})
export class PdsSelect {
  @Element() el!: HTMLPdsSelectElement;

  /**
   * Reference to the combobox wrapper
   */
  private comboWrapperRef?: HTMLDivElement;

  /**
   * Reference to the combobox input element
   */
  private comboInputRef?: HTMLDivElement;

  /**
   * Reference to the combobox overlay
   */
  private overlayEl: HTMLElement | null;

  /**
   * Track the index of the focused option
   */
  @Prop({mutable: true}) focusIndex = -1;

  /**
   * A unique identifier for the combobox
   */
  @Prop() componentId!: string;

  /**
   * Indicates that the combobox is disabled
   * @defaultValue false
   */
  @Prop() disabled = false;

  /**
   * Specifies the error text and provides an error-themed treatment to the field
   */
  @Prop() errorMessage?: string;

  /**
   * Displays a helper or description of the combobox
   */
  @Prop() helperMessage?: string;

  /**
   * Indicates  whether or not the input field is invalid or throws an error
   * @defaultValue false
   */
  @Prop() invalid = false;

  /**
   * Is enabled when the combobox is open
   * @defaultValue false
   */
  @Prop({ mutable: true }) isComboboxOpen = false;

  /**
   * Text to be displayed as the combobox label
   */
  @Prop() label?: string;

  /**
   * Determines the preferred position of the tooltip
   * @defaultValue "right"
   */
  @Prop({ reflect: true }) placement:
    'top'
    | 'top-start'
    | 'top-end'
    | 'right'
    | 'right-start'
    | 'right-end'
    | 'bottom'
    | 'bottom-start'
    | 'bottom-end'
    | 'left'
    | 'left-start'
    | 'left-end' = 'bottom-start';

  /**
   * The display id for the selected option
   */
  @Prop({mutable: true}) selectedOptionId?: string;

  /**
   * The display text for the selected option
   */
  @Prop({mutable: true}) selectedOptionText?: string;

  /**
   * The value for the selected option
   */
  @Prop({mutable: true}) selectedOptionValue?: string;

  /**
   * Emitted when the select value changes from selected option
   */
  @Event() pdsSelectChange!: EventEmitter<string>;

  componentWillLoad() {
    const options = Array.from(this.el.querySelectorAll('pds-select-option'));
    const firstSelectedOption = options.find((option) => option.hasAttribute('selected'));

    if (firstSelectedOption) {
      // Check if the 'selected' attribute exists and is not null

      if (firstSelectedOption.getAttribute('selected') !== null) {
        // Set the selected option value to the first selected option's value
        this.selectedOptionValue = firstSelectedOption.getAttribute('value');

        // Use innerHTML if the value is null
        if (this.selectedOptionValue === null) {
          this.selectedOptionText = firstSelectedOption.innerHTML;
        }
      }
    } else {
      // If no option is selected, get the first 'pds-select-option' and set the value accordingly
      const firstOption = this.el.querySelector('pds-select-option') as HTMLPdsSelectOptionElement;

      if (firstOption) {
        if(firstOption.innerHTML) {
          this.selectedOptionText = firstOption.innerHTML;
        }
      }
    }

    // Use shadowRoot to find the .pds-select__input element
    this.comboInputRef = this.el.shadowRoot?.querySelector('.pds-select__input') as HTMLDivElement;

    if (this.comboInputRef) {
      this.comboInputRef.addEventListener('click', this.handleComboboxToggle);
      // this.comboInputRef.addEventListener('blur', this.handleComboInputBlur);
    }
  }

  componentDidRender() {
    positionTooltip({elem: this.comboInputRef, elemPlacement: this.placement, overlay: this.overlayEl, offset:  4});
  }

  @Listen('pdsSelectOptionSelected')
  pdsSelectedOption(event: CustomEvent<any>) {
    const { id, text, value } = event.detail;

    // Set the value to equal the text if the value is empty
    if(this.selectedOptionValue === undefined) {
      this.selectedOptionValue = this.selectedOptionText;
    }

    if (this.selectedOptionValue !== value) {
      // Deselect the previous selected option, if any
      const prevSelectedOption = this.el.querySelector('pds-select-option[selected]') as HTMLPdsSelectOptionElement;

      if (prevSelectedOption) {
        prevSelectedOption.selected = false;
      }

      // Select the newly clicked option
      this.selectedOptionValue = value;
      this.selectedOptionText = text;

      const selectedOption = this.el.querySelector(`pds-select-option[component-id="${id}"]`) as HTMLPdsSelectOptionElement;

      if (selectedOption) {
        selectedOption.selected = true;

        // Update the selectedOptionId with the componentId of the selected option
        this.selectedOptionId = selectedOption.componentId;
      }

      this.pdsSelectChange.emit(this.selectedOptionValue);
    }

    this.handleComboboxToggle();
  }

  @Listen('keydown', {})
  handleComboInputKeyDown(event: KeyboardEvent) {
    switch (event.key || event.code) {
      // When the down arrow is pressed, open the combobox and focus the first option. The DOM focus remains on the this.isComboboxOpen
      // TODO: needs to be updated for down arrow. Should read... When the down arrow is pressed, open the combobox if it not already displayed without moving focus or chaning selection
      case 'ArrowDown':
        event.preventDefault();

        if (!this.isComboboxOpen) {
          this.handleComboboxToggle();
        }

        this.focusNextOption();
        break;
      
      // When the up arrow is pressed, open the combobox and focus the first option. The DOM focus remains on the this.isComboboxOpen
      case 'ArrowUp':
        event.preventDefault();

        if (!this.isComboboxOpen) {
          this.handleComboboxToggle();
          return
        }

        this.focusPreviousOption();
        break;
        
      case 'Escape':
        if(this.isComboboxOpen) {
          this.handleComboboxToggle();
        }
        break;

      // When the home key is pressed, open the combobox and moves visual focus to first selection
      case 'Home':
        event.preventDefault();
        if(!this.isComboboxOpen) {
          this.handleComboboxToggle();
        }
        this.focusFirstOption();
        break;

      // When the end key is pressed, open the combobox and moves visual focus to last selection
      case 'End':
        event.preventDefault();
        if(!this.isComboboxOpen) {
          this.handleComboboxToggle()
        }
        this.focusLastOption();
        break;

      // When the Enter key is pressed, open the combobox without moving focus or changing selection
      // When the Enter key is pressed, close the combobox, select the focused option, and move focus to the input
      case 'Enter':
        event.preventDefault();

        if (!this.isComboboxOpen) {
          this.handleComboboxToggle();
        }

        if (this.isComboboxOpen) {
          this.selectFocusedOption();
        }
        break;

      // When the SPACE key is pressed, open the combobox without moving focus or changing selection
      // When the SPACE key is pressed, close the combobox, select the focused option, and move focus to the input
      case ' ':
        event.preventDefault();

        // Open combobox if closed
        if (!this.isComboboxOpen) {
          this.handleComboboxToggle();
          return
        }

        // Select focused option if combobox is open
        if (this.isComboboxOpen) {
          this.selectFocusedOption();
        }
        break;
    }
  }

  private handleComboboxToggle = () => {
    this.comboWrapperRef =  this.el.shadowRoot?.querySelector('.pds-select') as HTMLDivElement;

    // update props
    this.isComboboxOpen = !this.isComboboxOpen;
    this.comboInputRef.setAttribute('aria-expanded', this.isComboboxOpen.toString());

    // open combobox
    if (this.isComboboxOpen) {
      this.comboWrapperRef.classList.add('is-open');

      // Move focus to the input when the combobox is opened
      this.comboInputRef?.focus();
      console.log('if this.isComboboxOpen');

      // Set the focus index to the first option
      // if (!this.focusIndex)  {
      //   console.log('IN HERE');
      //   this.focusIndex = 0;
      // }
    } else {
      // close combobox
      this.comboWrapperRef.classList.remove('is-open');

      // Reset focus index when the combobox is closed
      // this.focusIndex = -1;
    }
  };

  private selectInputClassNames() {
    const classNames = ['pds-select__input'];

    if (this.invalid && this.invalid === true) {
      classNames.push('is-invalid');
    }

    return classNames.join(' ');
  }

  // Focus Management
  private focusNextOption() {
    const options = this.el.querySelectorAll('pds-select-option');
    if (options.length > 0 && this.focusIndex < options.length - 1) {
      const previousIndex = this.focusIndex;
      this.focusIndex++;
      this.focusOptionAtIndex(previousIndex, this.focusIndex);
    }
  }

  private focusPreviousOption() {
    const options = this.el.querySelectorAll('pds-select-option');
    if (options.length > 0 && this.focusIndex > 0) {
      const previousIndex = this.focusIndex;
      this.focusIndex--;
      this.focusOptionAtIndex(previousIndex, this.focusIndex);
    }
  }

  private focusFirstOption() {
    const previousIndex = this.focusIndex;
    this.focusIndex = 0;
    this.focusOptionAtIndex(previousIndex, this.focusIndex);
  }

  private focusLastOption() {
    const options = this.el.querySelectorAll('pds-select-option');
    if (options.length > 0) {
      const previousIndex = this.focusIndex;
      this.focusIndex = options.length - 1;
      this.focusOptionAtIndex(previousIndex, this.focusIndex);
    }
  }

  // add previous index as well
  private focusOptionAtIndex(previousIndex: number, index: number) {
    console.log('focusOptionAtIndex: ', index);
    if (index === undefined) return false;

    const options = this.el.querySelectorAll('pds-select-option');

    // Update previousIndex
    if (previousIndex !== undefined && previousIndex < options.length) {
      // set to first option if combobox was initial opening
      if (previousIndex === -1){
        previousIndex = 0;
      }

      const prevOption = options[previousIndex].shadowRoot?.querySelector('.pds-select-option') as HTMLElement;
      prevOption.tabIndex = -1;
      prevOption.classList.remove('is--current');
    }

    // Update index
    if (index !== undefined && index < options.length) {
      const currentOption = options[index].shadowRoot?.querySelector('.pds-select-option') as HTMLElement;
      currentOption.tabIndex = 0;
      currentOption.classList.add('is--current');
      options[index].focus();
    }
  }

  // private selectFocusedOption() {
  //   const options = this.el.querySelectorAll('pds-select-option');
  //   if (options.length > 0 && this.focusIndex >= 0 && this.focusIndex < options.length) {
  //     const focusedOption = options[this.focusIndex].shadowRoot?.querySelector('.pds-select-option') as HTMLElement;
  //     focusedOption.click();
  //   }
  // }

  /**
   * Pending
   */
  @Method()
  async selectFocusedOption() {
    const options = this.el.querySelectorAll('pds-select-option');
    if (options.length > 0 && this.focusIndex >= 0 && this.focusIndex < options.length) {
      const focusedOption = options[this.focusIndex].shadowRoot?.querySelector('.pds-select-option') as HTMLElement;
      focusedOption.click();
    }
  }

  // End Focus Management

  render() {
    return (
      <Host>
        {this.label && (
          <label
            htmlFor={this.componentId}
            id={`${this.componentId}-label`}
            class="pds-select__label"
          >
            {this.label}
          </label>
        )}
        <div
          class={`
            pds-select

          `}
        >
          <div
            aria-controls={`${this.componentId}-listbox`}
            aria-describedby={assignDescription(this.componentId, this.invalid, this.helperMessage)}
            aria-expanded={this.isComboboxOpen.toString()}
            aria-haspopup="listbox"
            aria-labelledby={`${this.componentId}-label`}
            aria-activedescendant={this.selectedOptionId && this.isComboboxOpen ? `${this.componentId}-option-${this.selectedOptionId}` : undefined}
            class={this.selectInputClassNames()}
            id={this.componentId}
            ref={(el) => (this.comboInputRef = el)}
            role="combobox"
            tabindex="0"
          >
            {this.selectedOptionText || this.selectedOptionValue}
            <pds-icon name="caret-down" size="small"></pds-icon>
          </div>

          <div
            aria-labelledby={`${this.componentId}-label`}
            class="pds-select__menu"
            id={`${this.componentId}-listbox`}
            ref={(el) => (this.overlayEl = el)}
            role="listbox"
            tabindex="-1"
          >
            <slot></slot>
          </div>

          {this.helperMessage &&
            <p
              class="pds-select__helper-message"
              id={messageId(this.componentId, 'helper')}
            >
              {this.helperMessage}
            </p>
          }

          {this.errorMessage &&
            <p
              class="pds-select__error-message"
              id={messageId(this.componentId, 'error')}
              aria-live="assertive"
            >
              {this.errorMessage}
            </p>
          }
        </div>
      </Host>
    );
  }
}