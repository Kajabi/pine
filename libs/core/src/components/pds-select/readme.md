# pds-select



<!-- Auto Generated Below -->


## Properties

| Property                   | Attribute        | Description                                                                                                                                                       | Type                 | Default     |
| -------------------------- | ---------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------- | ----------- |
| `autocomplete`             | `autocomplete`   | Specifies if and how the browser provides `autocomplete` assistance for the field.                                                                                | `string`             | `undefined` |
| `componentId` _(required)_ | `component-id`   | A unique identifier used for the underlying component `id` attribute.                                                                                             | `string`             | `undefined` |
| `disabled`                 | `disabled`       | Indicates whether or not the select field is disabled.                                                                                                            | `boolean`            | `false`     |
| `errorMessage`             | `error-message`  | Displays error message text describing an invalid state.                                                                                                          | `string`             | `undefined` |
| `helperMessage`            | `helper-message` | Displays helper message text below select.                                                                                                                        | `string`             | `undefined` |
| `hideLabel`                | `hide-label`     | Visually hides the label text for instances where only the checkbox should be displayed. Label remains accessible to assistive technology such as screen readers. | `boolean`            | `undefined` |
| `invalid`                  | `invalid`        | Determines whether or not the select is invalid.                                                                                                                  | `boolean`            | `undefined` |
| `label`                    | `label`          | Text to be displayed as the select label.                                                                                                                         | `string`             | `undefined` |
| `multiple`                 | `multiple`       | Indicates whether multiple options can be selected.                                                                                                               | `boolean`            | `false`     |
| `name` _(required)_        | `name`           | Specifies the name. Submitted with the form name/value pair.                                                                                                      | `string`             | `undefined` |
| `required`                 | `required`       | Indicates whether or not the select field is required.                                                                                                            | `boolean`            | `false`     |
| `value`                    | `value`          | The value(s) of the selected option(s).                                                                                                                           | `string \| string[]` | `undefined` |


## Events

| Event             | Description                           | Type                      |
| ----------------- | ------------------------------------- | ------------------------- |
| `pdsSelectChange` | Emitted when a keyboard input occurs. | `CustomEvent<InputEvent>` |


## Shadow Parts

| Part       | Description |
| ---------- | ----------- |
| `"select"` |             |


## Dependencies

### Depends on

- pds-icon

### Graph
```mermaid
graph TD;
  pds-select --> pds-icon
  style pds-select fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


