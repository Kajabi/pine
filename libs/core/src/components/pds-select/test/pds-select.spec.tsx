import { newSpecPage } from '@stencil/core/testing';
import { PdsSelect } from '../pds-select';

describe('pds-select', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select component-id="field-1"></pds-select>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with label', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select label="Select Label"></pds-select>`,
    });
    if (page.root && page.root.shadowRoot) {
      const label = page.root.shadowRoot.querySelector('label');
      if (label) {
        expect(label.textContent).toBe('Select Label');
      } else {
        throw new Error('label is not available');
      }
    } else {
      throw new Error('page.root or page.root.shadowRoot is not available');
    }
  });

  it('renders with error message', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select component-id="field-1" error-message="Error occurred"></pds-select>`,
    });

    const errorMessage = page.root?.shadowRoot?.querySelector('.pds-select__error-message');
    expect(errorMessage?.textContent).toBe('Error occurred');
  });

  it('renders with helper message', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select component-id="field-1" helper-message="Helper text"></pds-select>`,
    });

    await page.waitForChanges(); // Ensures component fully renders

    const helperMessage = page.root?.shadowRoot?.querySelector('.pds-select__helper-message');
    expect(helperMessage?.textContent).toBe('Helper text');
  });

  it('renders with disabled attribute', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select component-id="field-1" disabled></pds-select>`,
    });

    await page.waitForChanges(); // Ensures component fully renders

    expect(page.root).toMatchSnapshot();
  });

  it('updates value prop on value change', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select options='[{"value": "paul", "label": "Paul McCartney"}, {"value": "john", "label": "John Lennon"}]' />`,
    });

    const pdsSelect = page.root;
    if (!pdsSelect || !pdsSelect.shadowRoot) {
      throw new Error('pdsSelect or pdsSelect.shadowRoot is not available');
    }
    const select = pdsSelect.shadowRoot.querySelector('select');

    // Change value to 'paul'
    if (select) {
      select.value = 'paul';
      select.dispatchEvent(new Event('change'));
      await page.waitForChanges();

      // Check if the value is updated
      expect(select.value).toBe('paul');

      // Change value to 'john'
      select.value = 'john';
      select.dispatchEvent(new Event('change'));
      await page.waitForChanges();

      // Check if the value is updated
      expect(select.value).toBe('john');

      // Change value to empty
      select.value = '';
      select.dispatchEvent(new Event('change'));
      await page.waitForChanges();

      // Check if the value is updated
      expect(select.value).toBe('');
    } else {
      throw new Error('select element is not available');
    }
  });

  it('parses valid JSON options correctly', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select options='[{"value": "1", "label": "Option 1"}, {"value": "2", "label": "Option 2"}]'></pds-select>`,
    });

    const pdsSelect = page.rootInstance;
    expect(pdsSelect.parsedOptions).toEqual([
      { value: '1', label: 'Option 1' },
      { value: '2', label: 'Option 2' },
    ]);
  });

  it('returns an empty array for invalid JSON options', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select options='invalid-json'></pds-select>`,
    });

    const pdsSelect = page.rootInstance;
    expect(pdsSelect.parsedOptions).toEqual([]);
  });

  it('returns an empty array for empty options', async () => {
    const page = await newSpecPage({
      components: [PdsSelect],
      html: `<pds-select options=''></pds-select>`,
    });

    const pdsSelect = page.rootInstance;
    expect(pdsSelect.parsedOptions).toEqual([]);
  });
});
