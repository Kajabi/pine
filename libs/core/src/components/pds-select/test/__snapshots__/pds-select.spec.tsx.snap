// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds-select renders 1`] = `
<pds-select component-id="field-1">
  <template shadowrootmode="open">
    <div class="pds-select">
      <label htmlfor="field-1"></label>
      <select class="pds-select__field" id="field-1"></select>
    </div>
  </template>
</pds-select>
`;

exports[`pds-select renders with disabled attribute 1`] = `
<pds-select aria-disabled="true" component-id="field-1" disabled="">
  <template shadowrootmode="open">
    <div class="pds-select">
      <label htmlfor="field-1"></label>
      <select class="pds-select__field" disabled="" id="field-1"></select>
    </div>
  </template>
</pds-select>
`;
